lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -147,6 +148,123 @@ void LocalEnforcer::setup(
+    SessionMap& session_map, const UpdateRequestsBySession& updates) {
+  for (auto& it : updates.requests_by_id) {
+    const std::string &imsi = it.first.first, &session_id = it.first.second;
+    SessionSearchCriteria criteria(imsi, IMSI_AND_SESSION_ID, session_id);
+    auto session_it = session_store_.find_session(session_map, criteria);
+    if (!session_it) {
+      MLOG(MWARNING) << "Not reporting session update event for " << session_id
+                     << " because it couldn't be found";
+      continue;
+    }
+    events_reporter_->session_updated(
+        session_id, (**session_it)->get_config(), it.second);
+  }
+}
+
+void LocalEnforcer::report_session_update_event_failure(
+    SessionMap& session_map, const UpdateRequestsBySession& failed_updates,
+    const std::string& failure_reason) {
+  for (auto& it : failed_updates.requests_by_id) {
+    const std::string &imsi = it.first.first, &session_id = it.first.second;
+    SessionSearchCriteria criteria(imsi, IMSI_AND_SESSION_ID, session_id);
+    auto session_it = session_store_.find_session(session_map, criteria);
+    if (!session_it) {
+      MLOG(MWARNING) << "Not reporting session update failure event for "
+                     << session_id << " because it couldn't be found";
+      continue;
+    }
+    std::ostringstream failure_stream;
+    failure_stream << "UpdateSession request to FeG/PolicyDB failed: "
+                   << failure_reason;
+    std::string failure_msg = failure_stream.str();
+    MLOG(MERROR) << failure_msg;
+    events_reporter_->session_update_failure(
+        session_id, (**session_it)->get_config(), it.second, failure_msg);
+  }
+}
+
+void LocalEnforcer::check_usage_for_reporting(
+    SessionMap session_map, SessionUpdate& session_uc) {
+  std::vector<std::unique_ptr<ServiceAction>> actions;
+  auto request = collect_updates(session_map, actions, session_uc);
+  execute_actions(session_map, actions, session_uc);
+  if (request.updates_size() == 0 && request.usage_monitors_size() == 0) {
+    auto update_success = session_store_.update_sessions(session_uc);
+    if (update_success) {
+      MLOG(MDEBUG) << "Succeeded in updating session after no reporting";
+    } else {
+      MLOG(MERROR) << "Failed in updating session after no reporting";
+    }
+    return;  // nothing to report
+  }
+
+  MLOG(MINFO) << "Sending " << request.updates_size()
+              << " charging updates and " << request.usage_monitors_size()
+              << " monitor updates to OCS and PCRF";
+
+  // set reporting flag for those sessions reporting
+  session_store_.set_and_save_reporting_flag(true, request, session_uc);
+
+  // Before reporting and returning control to the event loop, increment the
+  // request numbers stored for the sessions in SessionStore
+  session_store_.sync_request_numbers(session_uc);
+
+  // report to cloud
+  reporter_->report_updates(
+      request,
+      [this, request, session_uc,
```

Comment:
If this callback function is shared between here and LocalsessionManagerHandler, pull the callback function into a single function defined in `SessionReporter`. 
Always try to minimize the amount of code you have to copy :)